### PR TITLE
Fix issue where we weren't offering 'generate variable'

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
when there was an existing variable with the same name in an outer type.

Fixes https://github.com/dotnet/roslyn/issues/8358

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
retest vsi please

File: src/Features/CSharp/Portable/CodeFixes/GenerateVariable/GenerateVariableCodeFixProvider.cs [hunk 1]
```diff
@@ -24,11 +24,10 @@ internal class GenerateVariableCodeFixProvider : AbstractGenerateMemberCodeFixPr
         private const string CS0117 = nameof(CS0117); // error CS0117: 'TestNs.Program' does not contain a definition for 'blah'
         private const string CS0539 = nameof(CS0539); // error CS0539: 'Class.SomeProp' in explicit interface declaration is not a member of interface
         private const string CS0246 = nameof(CS0246); // error CS0246: The type or namespace name 'Version' could not be found
+        private const string CS0120 = nameof(CS0120); // error CS0120: An object reference is required for the non-static field, method, or property 'A'
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/GenerateVariable/GenerateVariableCodeFixProvider.cs]: Is this enough.  This means we will also offer to generate instance fields in the *current* class if there is already a static field, which doesn't seem right (and will conflict in name).

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/GenerateVariable/GenerateVariableCodeFixProvider.cs]: This just means the fixer *runs*.  It still does checking to make sure it is appropriate to add the fix.

This is yet another one of those cases where we have to exhaustively know every single error the compiler might emit when the thing you want is missing.  It's *super* annoying, and i really wish we didn't use this mechanism for code fix providers.

Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/GenerateVariable/GenerateVariableCodeFixProvider.cs]: And does it check for that case?

(BTW - I'm responsible for pushing for this design :tongue:)

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/GenerateVariable/GenerateVariableCodeFixProvider.cs]: THe design is unfortunate.  I don't mind tying us to the compiler's pass over the code to avoid having two systems churning over the world.  THe problem stems from us using diagnostics as the currency to try to back infer what we should do.  We really need to move to a mechanism that is more in line with our actual Compilation APIs (like GetSymbolInfo/GetTypeInfo).  For example, we could ask for a callback whenever GetSymbolInfo would return a specific candidate reason.  Right now we have an opaque "id + span" pair, and we work entirely off of that.  If the span isn't what we expect, things don't work.  If we dont' know about an ID, things don't work.  If the compiler suppresses a diagnostic, things don't work.  It's effectively moving us away from having a well designed and documented compiler API to use, to the moral equivalent of screen scraping.  It's not much better than us just runnign the compiler, seeing the emitted errors, and trying to go back from that to decide what to do.

I've discussed this issue with @srivatsn and @JohnHamby.  I really think the IOperation API should have a mechanism by which consumers can register actions to run *not* just on nodes, symbols, operations and tree, but *also* on strongly typed SymbolInfo information like 'CandidateReason'.  

In other words, given any node 'n', if calling semanticModel.GetSymbolInfo(n) would return a SymbolInfo 'si' with CandidateReason 'r', then i should be able to do something like "context.RegisterCandidateReason(r, callback)" and be called back with all relevant 'n's and 'si's.

This would allow us to move the IDE code back to what we had many years ago, with really clear and crisp code that used the actual compilation APIs, while still getting the perf benefit of running in tandem with the normal analysis pass.

**Comment by CyrusNajmabadi** [comment]
The case we're actually concerned about is the reverse.  Specifically, if we will offer to add a static field if the instance field already exists:

image

But we do not. 


In the case where you have a static field, there is no actual error (it's fine for an instance to write into a static):

image

**Comment by Pilchie** [comment]
Ah right.  Thank you.  One more thing to double check just to make me happy.  What if you reference an existing instance field through a type name explicitly:

```C#
class C
{
    int _field;
    void M()
    {
        C._field = 42;
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Not available there either :)

**Comment by Pilchie** [comment]
Okay :+1:

**Comment by CyrusNajmabadi** [comment]
I've added all of these as test cases.

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/4700/
// Retest reason: https://github.com/dotnet/roslyn/issues/9394
